src/model/Tick.ts

```diff
@@ -42,13 +58,37 @@ export class Tick {
     feeGrowthGlobal1X128: JSBI,
     leftToRight: boolean
   ): boolean {
-    // TODO
-    return false;
+    const liquidityGrossBefore = this.liquidityGross;
+    const liquidityGrossAfter = LiquidityMath.addDelta(
+      liquidityGrossBefore,
+      liquidityDelta
+    );
+    //assert(liquidityGrossAfter <= maxLiquidity)
+    const flipped =
+      JSBI.equal(liquidityGrossAfter, ZERO) !=
+      JSBI.equal(liquidityGrossBefore, ZERO);
+    if (JSBI.equal(liquidityGrossBefore, ZERO)) {
+      if (this.tickIndex <= tickCurrent) {
+        this._feeGrowthOutside0X128 = feeGrowthGlobal0X128;
+        this._feeGrowthOutside1X128 = feeGrowthGlobal1X128;
+      }
+    }
+    this._liquidityGross = liquidityGrossAfter;
+    this._liquidityNet = leftToRight
```

Comment:
I think logic here is reversed. Please see:
// when the lower (upper) tick is crossed left to right (right to left), liquidity must be added (removed)
1. upper - right to left - removed
2. lower - left to right - added

Reply:
Actually, here when handling the `upper` tick we always do as subtract, when handling the `lower` tick we always do as add. And then in swap process, if we move to a new tick, we add / subtract the tick's liquidity net to global liquidity according to swap direction (leftToRight / rightToLeft). This is important and please let me know if you have any questions.

Reply:
Good catch, the interface [here ](https://github.com/Uniswap/uniswap-v3-core/blob/b2c5555d696428c40c4b236069b3528b2317f3c1/contracts/libraries/Tick.sol#L146) is using `bool upper` which means right to left, and our interface is reversed, will modify.

Reply:
I see the point. It's me to blame that I changed the name of input paramater from `upper` to `leftToRight`. It's not proper, Let's talk about this tonight.